### ISSUE TITLE
Output preprocessed variables option

### ISSUE THREAD

**Issue by hgl** [postcss/postcss-custom-properties#22]
With this variables option:

```json
{
  "--border-color": "#aaa",
  "--background-color": "color(var(--border-color) alpha(0.2))"
}
```

It would be great if cssnext could give back a variables option that have all values preprocessed:

```json
{
  "--border-color": "#aaa",
  "--background-color": "rgba(170, 170, 170, 0.2)"
}
```
Otherwise it's very hard to consume the values for other js code.

I wonder what might be the best way to output this option? Should it be handled at cssnext level or this plugin here?

**Comment by hgl** [comment]
Also while testing this option, the `--` variable name prefix doesn't seem to be optional as the readme says. Removing them cause cssnext to complain the variables being undefined and needs a fallback.

**Comment by hgl** [comment]
After thinking about it again, I think the solution really shouldn't be preprocessing the passed variables option, but outputting the final computed custom properties, no matter if a variables option is passed or not.

In other words:

```css
:root {
  color: #aaa;
}
body {
  color: var(--color);
}
```
Preprocessing this should output the final css along with the computed custom properties, if an option is passed:

```
$ cssnext --custom-properties props.json file
```
props.json
```json
{
  "color": "#aaa"
}
```

Not sure if it's doable with postcss?

**Comment by MoOx** [comment]
The plugin cannot return anything but just edit the CSS AST.
Maybe we can offer a mutable option but I think we should instead just compute JS variables like it's done for others properties.
PRs are welcome for this enhancement.

For the optional --, just check this commit that have a test in it https://github.com/postcss/postcss-custom-properties/commit/0691784ed2218d7e6b16da8c4df03e2ca0c4798c

**Comment by hgl** [comment]
> but I think we should instead just compute JS variables like it's done for others properties.
I'm sorry but not quite sure what do you mean by this. Could you elaborate a bit?

> For the optional --, just check this commit that have a test in it 0691784
I was running an older version of this plugin. Now it's working. Thanks for the heads up.

**Comment by MoOx** [comment]
I mean resolve var() in JS variables. Should be pretty easy.

**Comment by hgl** [comment]
OK, do you mean we should resolve the passed in variables option directly without looking at the actual css source code?

This can fail if users do:

variables option
```json
{ "border-color": "var(color)" }
```
css
```css
:root {
  "color": "#aaa";
}
div {
  border: 1px solid var(--border-color);
}
```
Without the css source code, the passed variables option might be incomplete.

I think the mutable option is a viable approach, what if we modify the `variables` option in place? 

```js
var opts = {};
postcss()
  .use(customProperties(opts))
  .process(css);
opts.variables // computed values
```

and then cssnext could use this property to output the result.

I can starting working a PR if you think this approach is good.

**Comment by MoOx** [comment]
This plugin works with 2 passes on the complete AST. If we try to resolve JS variables between the first resolution and the variables replacement, I think we should be good. Just here https://github.com/postcss/postcss-custom-properties/blob/0691784ed2218d7e6b16da8c4df03e2ca0c4798c/index.js#L85. Should take a second.
I am not a fan of mutability :)

**Comment by hgl** [comment]
So after the JS variables are resolved, how do we output them?

Are you saying we let cssnext operate directly on json files?

variables.json
```json
{ "border-color": "var(color)" }
```
style.css
```css
:root {
  "color": "#aaa";
}
div {
  border: 1px solid var(--border-color);
}
```
```
$ cssnext --compute-js-variables variables.json style.css

{ "border-color": "#aaa" }
```

**Comment by MoOx** [comment]
We are not in cssnext. This plugin will never handle file.

I just pushed a modification that should make you happy ;)
I won't do anything more. Your first example in #22 with color won't be resolved by this plugin.
If you want to play with color, you can do that before offering JS variables... using JS ! :p

**Comment by MoOx** [comment]
Closed by a5aa2bfe92d7b5414a9037efa3da95609f2ffdff

**Comment by hgl** [comment]
OK. I see where our misunderstanding lies.

The bug you fixed actually doesn't exist. 

```json
{
  "--border-color": "#aaa",
  "--background-color": "var(--border-color)"
}
```

Using this variables option works fine before your fix. Because at the second resolution, the js variable will be [recursively resolved](https://github.com/postcss/postcss-custom-properties/blob/78e550eb0e60c00f7430c70f5285872c2c8d1a40/index.js#L71).

My feature request was to output the resolved values of all custom properties. I [create a commit at my branch](https://github.com/hgl/cssnext/commit/7e1b333b0e663be27880a75f6009bcb41eac4ae3). It currently only works with custom properties specified directly in the source code and whose values don't contain `var()`. To make it work with custom properties specified with `variables` option and those containing `var()`, this plugin needs to offer an extra feature (we can discuss it once you have seen my commit).

Let me know if you want such feature in cssnext.

(And if I'm not wrong, this fix should be reversed.)

**Comment by MoOx** [comment]
I will see if what I did was stupid.
I don't know what you are trying to do, but you have weird needs. You should defined JS vars and use them as is... Don't make things complicated...

```js
var color = "#aaa"
var props = {
  "--border-color": color,
  "--background-color": "color(" + color + " alpha(0.2))"
}
```

FYI https://github.com/harthur/color/issues/55

**Comment by hgl** [comment]
> but you have weird needs

I need to use these variables in my app, for doing UI animations with [web-animations](https://github.com/web-animations/web-animations-next) for example:

```js
var startColor = variables["start-color"];
var endColor = variables["end-color"];
elem.animate([color: startColor, endColor], 3000);
```
```json
{
  "--start-color": "#aaa",
  "--end-color" : "color(var(--start-color) alpha(0.2))"
}
```
```css
a {
  color: var(--start-color);
}
a:hover {
  color: var(--end-color);
}
```
The same custom properties will be used both in css and js. So I need the resolved value in order to make it work in most browsers.

**Comment by hgl** [comment]
So what do you think? Could cssnext offer a list of resolved custom properties?

**Comment by MoOx** [comment]
As I said before, for me you can do that in JS first, without adding an extra step.
If you want to share same variables from JS & CSS, just prepare your variables in JS first.

In your case you could do

```js
var color = require("color")
var mainColor = color("#aaa")

module.exports = {
  "start-color": mainColor.rgbString(),
  "end-color" : mainColor.alpha(0.2).rgbString()
}
```
cssnext is in javascript so everything that is implemented come from JS. You can use JS directly.

**Comment by MoOx** [comment]
Note: that what me and other people are doing.

**Comment by hgl** [comment]
Thanks for showing me the tip. Didn't realize this.

However, not super excited about this procedural approach. I'm worried the number of boilerplate code it entails:

```js
var color = require("color")
var mainColor = color("#aaa")
var pageWidth = 960
var sidebarWidth = pageWidth / 2

module.exports = {
  "start-color": mainColor.rgbString(),
  "end-color" : mainColor.alpha(0.2).rgbString(),
  "page-width": pageWidth + "px",
  "sidebar-width": sidebarWidth + "px",
  "search-width": sidebarWidth * 0.8 + "px"
}
```

Comparing it with
```json
{
  "start-color": "#aaa",
  "end-color" : "color(var(--start-color) alpha(0.2))",
  "page-width": "960px",
  "sidebar-width": "calc(var(--page-width) / 2)",
  "search-width": "calc(var(--sidebar-width) * 0.8)"
}
```

I personally find the later less painful to write.

But I understand if you don't want the extraction feature in cssnext, and I'm thinking about writing an separate project to provide this feature, but I'd still like to leverage this project to resolve variables. Will you accept an extra option where custom-properties are preserved but their values are replaced with resolved ones?

**Comment by MoOx** [comment]
If you do a correct PR with tests & docs, why not.
But what you are trying to looks like a dangerous game.

This es6 code looks better to me, and more flexible:

```js
var color = require("color")

const mainColor = color("#aaa")
const pageWidth = 960

exports default {
  "start-color": mainColor.rgbString(),
  "end-color" : mainColor.alpha(0.2).rgbString(),
  "page-width": `${pageWidth}px`,
  "sidebar-width": `${pageWidth / 2}px`,
  "search-width": `${pageWidth * 0.8}px`
}
```

It's explicit.

**Comment by hgl** [comment]
`search-width` depends on `sidebar-width`, so it's actually 
```
"search-width": `${pageWidth / 2 * 0.8}px`
```

Will send a PR soon.

**Comment by hgl** [comment]
Ideally this extra option should extends the existing `preserve` option:

```
preserve: "specified"
preserve: "computed"
preserve: "both"
```

Since the feature I'm looking overlaps with the existing `preserve`. `true` can be aliased to `both` to provide backwards compatibility. What do you think? Does `preserve: "specified"` make sense?

**Comment by MoOx** [comment]
`preserve` is just a way to keep original output.
What you are trying to do is a complete hack since it's rely on the fact that all plugins are chained and resolve each of their responsibility (postcss-color-*, postcss-calc, etc..).

I think we can't accept such an option in this plugin since it's beyond its responsibility.

**Comment by hgl** [comment]
> since it's rely on the fact that all plugins are chained

That's the reason I'm currently implement this in cssnext. From cssnext point of view, relying on plugins being chained is not hack, and the separate project will mostly copy cssnext, but using only postcss plugins that can affect css properties values.

The feature I'm proposing to this plugin is that, comparing to the current preserve behavior:

```css
:root {
  --width: 100px;
  --height: var(--width);
}
/* compiles to */
:root {
  --width: 100px;
  --height: 100px;
  --height: var(--width);
}
```
It will do:

```css
:root {
  --width: 100px;
  --height: var(--width);
}
/* compiles to */
:root {
  --width: 100px;
  --height: 100px;
}
```

That is, it will not leave the specified value. Instead, it replaces it with the resolved value. So at cssnext level, I can retrieve all custom properties without get duplicate values.

I don't think this proposal is a hack. It doesn't rely on other plugins. It [simply removes the specified declaration without the conditional](https://github.com/postcss/postcss-custom-properties/blob/master/index.js#L178) as the code base currently does.

Hope that will change your mind.

**Comment by MoOx** [comment]
So what about a simple other option `preserveComputedOnly` ?
Or maybe: 
preserve: true || "all" (current behavior)
preserve: "original" // only preserve original authored values
preserve: "computed" // only preserve computed values

**Comment by hgl** [comment]
I'm incline to the later, otherwise we need make a decision if `preserve` overrides `preserveComputedOnly` or the other way around.

I was going to suggest "original", but in css spec, they use ["specified"](http://dev.w3.org/csswg/css2/cascade.html#specified-value). So I thought it might be nice to follow suite.

Let me know if you insist.

**Comment by MoOx** [comment]
if specs say specified go for this 

preserve: true || "all" (current behavior)
preserve: "specified" // only preserve original authored values
preserve: "computed" // only preserve computed values

**Comment by hgl** [comment]
Cool. Glad we have an agreement. :)

**Comment by hgl** [comment]
Should I work on the 3.2.0 commit?

**Comment by MoOx** [comment]
Please just remove useless lines in the module BUT keep the tests to ensure we are not going back :)
Don't drop any commit, just iterate.

**Comment by MoOx** [comment]
The thing is you think about declaration. `preserve` affect declaration + var() usage.
So indeed this should be moved to another option.
preserve will affect the "computed" option. Keep that in mind.

**Comment by hgl** [comment]
source
```css
:root {
	--color: #aaa;
	--bg-color: var(color);
}
body {
	background: var(--bg-color);
}
```
`{preserve: true, computed: "add"}`

```css
:root {
	--color: #aaa;
	--bg-color: #aaa;
	--bg-color: var(color);
}
body {
	background: #aaa;
	background: var(--bg-color);
}
```
`{preserve: true, computed: "replace"}`
```css
:root {
	--color: #aaa;
	--bg-color: #aaa;
}
body {
	background: #aaa;
}
```
`{preserve: false, computed: "add"}`
```css
body {
	background: #aaa;
	background: var(--bg-color);
}
```
`{preserve: false, computed: "replace"}`
```css
body {
	background: #aaa;
}
```

I think `computed` should be dominant over `preserve`, because as can be seen from the examples, `{preserve: false, computed: "add"}` doesn't make much sense.

What about we default `computed` to `replace`, `preserve` to `false`, when `computed` is `add`, `preserve` is simply ignored and works like `true`? This make the plugin backward compatible and won't generate incorrect css code.

**Comment by MoOx** [comment]
I was thinking `computed` was only going to affect custom props definition.
Imo, this option should just not be enabled by default, so no issue for backward compatibility.
Maybe you can simply add an option like `computedProperties` that will only handle --* definition ?

**Comment by hgl** [comment]
If we introduce an option only affects custom props definition:

`{preserve: true, computedProperties: true}`

```css
:root {
    --color: #aaa;
    --bg-color: #aaa;
    --bg-color: var(color);
}
body {
    background: #aaa;
    background: var(--bg-color);
}
```

`{preserve: true, computedProperties: false}`
```css
:root {
    --color: #aaa;
    --bg-color: #aaa;
}
body {
    background: #aaa;
    background: var(--bg-color);
}
```

This feels asymmetrical (also the name `computedProperties` doesn't feel right in these examples.). If we keep `background: var(--bg-color);` so it can override `background: #aaa;` in supporting browsers, why don't we keep `--bg-color: var(color);` so it can override `--bg-color: #aaa;`?

**Comment by hgl** [comment]
What about `preserveSpecifiedDefinition`? Doesn't work when preserve is `false`, defaults to `true` when `preserve` is true.

**Comment by hgl** [comment]
I just realized that at cssnext level, i'm actually able to dedup the custom properties definitions (when there are multiple custom properties definitions with the name, use the last one doesn't contain `var()`). So a `preserve: true` offer by this plugin is enough to achieve the goal.

But I also realized I need the same preserve feature from the custom media plugin (want to extract custom media queries also.)

Will write a doc to clearly explain the problem and my plan of solving it. Sorry for all the noises I generated.
